Stale events support

paolodamico opened the pull request:
## Changes
Closes #4791.
- Implements `created_at` & `last_seen_at` to `EventDefinition`
- Created at is set automatically now when the object is created. We also have a migration to backfill old objects.
- `last_seen_at` is updated from the plugin server every ~15 minutes. We use caching in the plugin server to avoid adding too much overhead on Postgres DB.


<img width="671" alt="" src="https://user-images.githubusercontent.com/5864173/144332189-39e990dc-c922-4a56-aeb1-aa8d07f1ce86.png">

<img width="635" alt="" src="https://user-images.githubusercontent.com/5864173/144332199-85f202e0-f9bf-4945-a47c-3b5982c6a3e3.png">

## How did you test this code?
- Tested with demo data. Old events that didn't have `last_seen` showed the stale indicator. New events did not.
- For the migration, tested with Postgres. Ran migrations up to 0185, then `python manage.py setup_dev`, then ran migration 0186 and ensured that `created_at` was properly populated.
- Functional tests for plugin server & API response.

paolodamico committed:
add created_at & last_seen_at

paolodamico committed:
update last_seen_at on plugin server

paolodamico committed:
fix tests

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -55,11 +61,17 @@ export class TeamManager {
                 'insertEventDefinition'
             )
             this.eventNamesCache.get(team.id)?.add(event)
+        } else {
```
o_o this is likely not a good idea - great way to create even more load for the postgres database and cause us to fall over.

Note that this is done every single event ingestion - it really doesn't need to be that frequent.

paolodamico replied:
Thanks for the early feedback, don't want to spend any more time going down the wrong direction. That was my first thought but I didn't want to add another background job that requires going through the entire list of events as we've had problems with that in the past (though it could be greatly optimised). 

How about relying on the caching mechanism we have to see if the event definition has been created and only update when 30min-1hr has passed?

mariusandra replied:
I guess the true solution is replacing `ASYNC_EVENT_PROPERTY_USAGE` with a plugin-server-based counter that updates `volume_30_day` on the fly, syncing data to postgres every 10-60 sec or so.

paolodamico replied:
I'll keep the `last_seen_at` as based on @clarkus's feedback, it'll be helpful to tell this information to the user (which `volume_30_day` wouldn't provide. I have an idea on how to improve the `ASYNC_EVENT_PROPERTY_USAGE` method, but for a separate PR.

macobo replied:
Syill think this is a bad idea - pretty sure this will blow up ingestion one way or another sooner or later - either queries or memory wise. 

Gonna cc @yakkomajuri here.

paolodamico replied:
Thanks for the pushback here, will wait for @yakkomajuri's input before moving forward.

paolodamico replied:
An alternative to prevent blowing up the DB is extending the threshold, maybe just update once a day instead of every 15 minutes, but unsure about memory. I'm worried that doing it via an async job will be worse, we'd have to go almost through all possible event names.

macobo replied:
If you `SELECT DISTINCT (team_id, event) FROM events` once per day in a cronjob + insert the same timestamp for all rows the query wouldn't be too bad.

yakkomajuri replied:
Some thoughts (trying to limit the amount of time I brainstorm on this):

1. Agree with @macobo, let's be careful with ingestion-triggered Postgres queries, particularly those to supplement functionality that isn't core to ingestion (as opposed to creating persons and distinct IDs for example).
2. A cronjob sounds like a it could be a good approach to me. Also, as you're planning to backfill, you'd only need to look through the events seen on that given day (use `_timestamp` for that), so might not be that bad. 
3. There could potentially be another solution here if you're feeling adventurous. You could leverage CH materialized views to do this. Although in that case you couldn't pull all the data you need from the event definition Postgres row and would have to hit CH.

I'd say probably test out what a daily cronjob would look like and see how it might perform. It could also help with performance to remove granularity here and store `last_seen_at` to the "nearest day", so you don't need to pull timestamps and instead just do something like:

```
SELECT DISTINCT (team_id, event) FROM events WHERE _timestamp > today()
```

Final point is that it could be worth starting this out as an enterprise feature to test it out.

paolodamico committed:
cache last_seen_at

paolodamico committed:
existing tests now work

paolodamico committed:
cache last_seen_at @ launch + more tests

paolodamico committed:
fix additional tests

paolodamico committed:
update migration

paolodamico committed:
Merge branch 'master' into stale-events

paolodamico committed:
prefill created_at

paolodamico committed:
add test

paolodamico committed:
fix migration for FOSS

paolodamico committed:
stale indicator

paolodamico committed:
final tweaks

paolodamico committed:
fix mypy

paolodamico commented on frontend/src/lib/components/TaxonomicFilter/InfiniteList.tsx:
```diff
@@ -41,19 +45,44 @@ export function tooltipDesiredState(element?: Element | null): ListTooltip {
 }): JSX.Element | string => {
+    const parsedLastSeen = (item as EventDefinition).last_seen_at ? dayjs((item as EventDefinition).last_seen_at) : null
+    const isStale =
+        (listGroupType === TaxonomicFilterGroupType.Events && !parsedLastSeen) ||
```
I'm not 100% sure of `!parsedLastSeen`, it means that for a while this stale indicator may throw some false positives, where an event hasn't been received since we push this change, but removing it means that very old events that are no longer sent will never be marked as stale.

mariusandra replied:
Perhaps remove the "and looks like is no longer used" part if there's no clear information on its usage? Probably OK to still show "stale" then...

paolodamico commented on frontend/src/scenes/experiments/experimentsLogic.tsx:
```diff
@@ -3,6 +3,7 @@ import { api } from 'lib/api.mock'
 import { teamLogic } from 'scenes/teamLogic'
 import { Experiment } from '~/types'
 
+import { experimentsLogicType } from './experimentsLogicType'
```
This was added automatically, looks like it's right?

paolodamico commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -55,11 +63,26 @@ export class TeamManager {
+            if (
+                !this.eventLastSeenCache.get(`${team.id}_${event}`) ||
+                eventTimestamp.diff(
+                    this.eventLastSeenCache.get(`${team.id}_${event}`) || DateTime.fromISO('1970-01-01')
```
This looks dirty, unsure if there's a better way to say DateTime.min?

paolodamico commented on posthog/migrations/0186_stale_events.py:
```diff
@@ -0,0 +1,60 @@
+from posthog.constants import AnalyticsDBMS
+
+
+def set_created_at(apps, schema_editor):
```
Unsure whether this backfill is a good idea, it could potentially take a long time (though I don't think it'd freeze the table). I could also use some help testing the CH version as I have problems running it on my computer (opted not to waste time fixing it as the M1 is coming pretty soon).

mariusandra reviewed (commented):
Can I suggest a different approach. I think this fixes the performance issue and remains maintainable.

```ts
// somewhere
let eventCache = new Map<string, number>()

// when ingesting
eventCache.set(`${team_id}::${event_name}`,new Date().valueOf())
if (eventCache.size > 100000) { // to not run out of memory
  flush()
}

// somewhere else, or we could have `if timeSinceLastFlush > 60 then flush` above
const nextFlush = () => { setTimeout(() => { flush(); nextFlush() }, 60000) }
nextFlush()

// on server quit
also call flush()

// the flush:
function flush() {
    const events = eventCache
    eventCache = new Map()
    // make a batch of postgres queries that for each row in the cache:
    // "insert into event_definition (X,Y,Z) ON CONFLICT update event_definition X=Y"
}
```

This would make just one (large) one-way query to postgres with a bunch of update statements, once every minute per thread.

I'd also add `volume_total`, and `+=` it to have a near real-time counter.

paolodamico committed:
update with Marius approach - tests TBC

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -42,6 +44,29 @@ export class TeamManager {
+
+        for (const event of this.eventLastSeenCache) {
+            team_ids.push(event[0].split('_')[0])
+            event_names.push(event[0].split('_')[1])
```
This doesn't work - event names _will_ have any character in it, so this will have bugs.

paolodamico replied:
I've already changed this to just grab the first `_`

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -42,6 +44,29 @@ export class TeamManager {
+            `UPDATE posthog_eventdefinition t1 SET t1.last_seen_at = GREATEST(t1.last_seen_at, t2.last_seen_at)
+            FROM (UNNEST ($1) as team_id, UNNEST($2) as name, UNNEST($3) as last_seen_at) as t2
+            WHERE t1.name = t2.name AND t1.team_id = t2.team_id`,
+            [team_ids, event_names, last_seen_at_array],
```
This was brought up elsewhere, but why do we want track everything so precisely?

I don't see a business reason why you even need hourly precision here not to mention second-wise, and it makes everything perform worse. In this case we increase our memory usage and need to send more bytes over to postgres for the update.

It's also one of these decisions that are really hard to undo after the fast. I'd push really hard for using daily resolution

paolodamico replied:
Agreed! From a product perspective having daily resolution (`Date` instead of `DateTime` is enough). If we do move forward with this approach I'll update this to use only dates.

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -42,6 +44,29 @@ export class TeamManager {
+        const event_names: string[] = []
+        const last_seen_at_array: number[] = []
+
+        for (const event of this.eventLastSeenCache) {
```
I think this is not safe for concurrent modifications, which will happen here. You want to do:

```
const events = this.eventLastSeenCache
this.eventLastSeenCache = new Map()
```

just above here.

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -63,25 +88,22 @@ export class TeamManager {
+            }
+            if (this.eventLastSeenCache.size > 100000 || DateTime.now().diff(this.lastFlushAt).minutes > 60) {
+                // to not run out of memory
+                void this.flushLastSeenAtCache()
```
You should await here. Voided promises are most often _not_ flushed properly.

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -63,25 +88,22 @@ export class TeamManager {
+            if ((this.eventLastSeenCache.get(`${team.id}_${event}`) ?? 0) < DateTime.now().valueOf()) {
+                this.eventLastSeenCache.set(`${team.id}_${event}`, eventTimestamp.valueOf())
+            }
+            if (this.eventLastSeenCache.size > 100000 || DateTime.now().diff(this.lastFlushAt).minutes > 60) {
```
Q: Why did we end up going with this approach? IIRC everyone was pushing towards using a cronjob if possible.

macobo commented:
For context why I'm trying to push this out of the plugin-server:
1. Over the past year I remember 2 releases that were broken due to similar logic, where users couldn't ingest events due to overloading one service or the other.
2. As we scale, scaling plugin-server and especially postgres is hard, we _will_ be spending man-years building towards a more scalable solution. That's really expensive and delaying this work is cruicial for the sake of the company. This means _avoiding_ doing unneeded work and being clever about work that is done.

For reference, I don't think the event/property stats stuff should be in plugin server in the first place and will be one of the first things we'll move off of it as we scale.

paolodamico committed:
add more tests

paolodamico committed:
address feedback

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -39,7 +44,38 @@ export class TeamManager {
+
+        for (const event of events) {
+            const [key, value] = event
+            team_ids.push(key.split('_')[0])
```
Suggestion:

```
team_id, event = key.split('_', 1)
```

paolodamico commented:
From my conversation with @macobo, I'll pause my work on this and wait for a final resolution (this approach is mostly done if we do move forward with it). I don't have all the context required to make the right call here, so deferring to @mariusandra / @yakkomajuri.

paolodamico committed:
apply suggestion

macobo commented:
Note from the discussion we had with Paolo:
- We relaxed the time requirements here to only refresh last_seen every day or so based on current time (not event time). Refreshing date every day works business-logic wise here and makes calculations/updates perform better both in plugin-server and in async solutions.
- There are potential other approaches suggested, which might become easier to implement thanks to the additional constraint. It's up to Paolo to decide whether to go with this solution or those ones.

paolodamico committed:
static dates that don't change with computer timezone

paolodamico committed:
only enable for specific teams

paolodamico committed:
tests for temporary env var

paolodamico committed:
finally get update statement right

paolodamico committed:
full tests

paolodamico committed:
Merge branch 'master' into stale-events

paolodamico committed:
fix merge conflict

paolodamico commented:
Alright, ended up spending hours making this work but it's finally there. Based on the latest conversation I had with @mariusandra, this is ready for a final review and we can ideally merge tomorrow and test out with our team only. Latest updates:
- Added missing tests.
- Introduced env var so that feature is only enabled for certain instances and for specific teams only.
- Merge conflicts & rebase.
- Based on our last conversation we talked about removing only milliseconds from granularity of `last_seen_at` timestamp (as opposed to fully moving to `Date`). You have better context on the future use case for this, but I opted not to do this as we gain almost zero benefits from dropping just ms (values are still sent to Postgres and dropped there). We could still do the full jump to date if it makes sense.

paolodamico committed:
remove duplicate import

mariusandra commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -12,15 +14,22 @@ export class TeamManager {
         this.eventPropertiesCache = new Map()
         this.instanceSiteUrl = instanceSiteUrl || 'unknown'
+        this.lastFlushAt = DateTime.now()
+        this.experimentalLastSeenAtEnabledTeams = experimentalLastSeenAtEnabledTeams?.split(',') ?? []
```
nit: this could be sanitized to numbers directly here

mariusandra commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -39,7 +48,45 @@ export class TeamManager {
+        for (const event of events) {
+            const [key, value] = event
+            const teamId = key.split('_', 1)[0]
+            const eventName = key.substring(key.indexOf(teamId) + teamId.length + 1)
```
This part is confusing and it's not obvious that `split('_', 2)` wouldn't work. I'd either add a comment explaining why `substring` needs to be used, or replace it with something cleaner like:

```
// split('_', 2) eats chunks 3 and up, so we must do this:
const [teamId, ...rest] = key.split('_')
const eventName = rest.join('_')
```

mariusandra replied:
The other option is to use as a key: `JSON.stringify([teamId, eventName])`... and read it back with `const [teamId, eventName] = JSON.parse(key)`

paolodamico replied:
Like the JSON idea, done!

mariusandra commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -39,7 +48,45 @@ export class TeamManager {
+        if (params.length) {
+            await this.db.postgresQuery(
+                `UPDATE posthog_eventdefinition AS t1 SET last_seen_at = GREATEST(t1.last_seen_at, to_timestamp(t2.last_seen_at::integer))
+                FROM (VALUES ${valuesStatement.substring(
+                    0,
+                    valuesStatement.length - 1
+                )}) AS t2(team_id, name, last_seen_at)
```
This `valuesStatement.substring(0,valuesStatement.length - 1)` feels like something can go horribly wrong. Just as a practice we should almost never really inject sql into sql, and removing a comma like that is... smelly 🐟 

It's also not clear why `from (values())`, and not just something like:

```sql
UPDATE posthog_eventdefinition SET last_seen_at = GREATEST(last_seen_at, $1), created_at = LOWEST(created_at, $2) where team_id = $3 and name = $4
```

Finally, I'd check about running all of this in a transaction. I can't 100% confirm this, but I think a transaction will be sent to postgres in one go. So if we have 1000 events to update, here we're awaiting for 1000 queries to return, so combining them all into one block that's sent to the backend at the same time will be faster.

paolodamico replied:
Alright, so the reason this uses `from (values())` and we inject SQL (which doesn't contain any user input so should be clean), is to run the update statement in one go, it's a single query. I'll switch to transaction-based updates in my last commit here, so you can compare and advise on which option would be best.

mariusandra commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -54,12 +101,37 @@ export class TeamManager {
+                if ((this.eventLastSeenCache.get(`${team.id}_${event}`) ?? 0) < eventTimestamp.valueOf()) {
+                    this.eventLastSeenCache.set(`${team.id}_${event}`, eventTimestamp.valueOf())
+                }
+                if (this.eventLastSeenCache.size > 100000 || DateTime.now().diff(this.lastFlushAt).minutes > 360) {
+                    // to not run out of memory
+                    await this.flushLastSeenAtCache()
+                }
```
Running this every 6 hours or one hundred thousand unique event types is... much.

Can these numbers also be configurable by envs? I'd also try them with much lower values. E.g. 1k events and 1minute, especially if just rolling out to our team.

Finally, I'd also liberally add "status.log"-s everywhere to be able to see from logs what's happening. E.g. if we're now flushing because of the timeout or the memory limit.

paolodamico replied:
Lowering the values, and adding logs everywhere, but won't do the env var at this time. As we're only testing with our team, it's the same effort / impact to change an env var than the code (ECS task definition) if we need to adjust while testing. If we move forward with this, I'll add the env var.

mariusandra commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -39,7 +48,45 @@ export class TeamManager {
     }
 
-    public async updateEventNamesAndProperties(teamId: number, event: string, properties: Properties): Promise<void> {
+    async flushLastSeenAtCache(): Promise<void> {
```
Can we log this to the console, together with e.g. nr of events, total elapsed time for the flush, etc.

paolodamico replied:
Yessir

mariusandra reviewed (changes requested):
Left some thoughts inside. Overall I think it's worth trying out, but still a few things I'd improve.

paolodamico committed:
Merge branch 'master' into stale-events

paolodamico committed:
clean up cache keys

paolodamico committed:
better logging, reduce processing time

paolodamico committed:
address final feedback

paolodamico committed:
alt approach: use psql transactions

paolodamico commented:
All feedback addressed, ready to 🚢 ?

paolodamico committed:
fix SAML deps on CI?

paolodamico committed:
missing updates

paolodamico committed:
fix ingest event tests

mariusandra reviewed (commented):
Final piece of feedback:

1. I'd add in addition to the console logging, some logging to kafka, a'la:
```
    server.statsd?.timing('kafka_queue.each_event', eachEventStartTimer)
    server.internalMetrics?.incr('$$plugin_server_events_processed')
```

2. Will this suddenly show all events as stale? Perhaps let's put it behind a flag to avoid extra support work? We could build a backfill later as well.

3. We are entering uncharted waters here. While this behaves fine locally, it may behave totally strange in production. So we need to release gradually, run experiments, and see how much time things take in grafana. It's very possible your original query will be much faster than this, understanding it better. So I'd bring it back actually --> it is still one statement... or then have some way of measuring between the two approaches via grafana.

I guess we can't really use feature flags or experiments here to live toggle things?

paolodamico committed:
Revert "alt approach: use psql transactions"

This reverts commit 70d15b5a638cd05613001c9b8a6d561a5bc2bb1d.

paolodamico committed:
add statsd info

paolodamico committed:
FF indicator

paolodamico committed:
prettier

paolodamico committed:
eslint

paolodamico committed:
it's hard to make small changes without a dev env

mariusandra committed:
fix luxon timing

mariusandra committed:
divide by 1000

mariusandra committed:
flush cache in all workers

mariusandra reviewed (approved):
Found three issues. All fixed now.

1. The `.seconds` and `.minutes` diff helpers were always returning 0. Turns out the object behind the scenes was `{ milliseconds: 13336 }` and you need to specify `.as('seconds')` instead. Fixed it.

2. Then, the insert query was broken. On the 7th day of Christmas, Postgres said to me:

```
[___1] 23:08:22 🔔 error: value "1638914888189" is out of range for type integer
```

Which it indeed is. Dividing the event key by 1000 and calling it a `numeric` solved the problem.

3. Finally, `closeJobs` will only run the flush on the main thread. We need to run it on all worker threads. Patched that as well. However we're never 100% sure if this runs always. The server can just shut off. So it's always good to flush frequently.

No idea if tests will pass 🤞. 

I'm going to give it the ✅, but only merge it after tests pass, and monitor after that all is good :).

mariusandra reviewed (changes requested):
That's a weird lot of cypress errors, all connected with the filter. So going to change to the vote.

paolodamico commented:
Sad, alright I'll fix the Cypress tests once I get back to my old computer and we can merge tomorrow. @fuziontech also expressed some strong interest in suggesting some improvements before merging this.

paolodamico committed:
fix cypress

paolodamico committed:
Merge branch 'master' into stale-events
